crates/codegen/ebnf/src/builder.rs

```diff
@@ -0,0 +1,432 @@
+use std::collections::HashMap;
+
+use codegen_language_definition::model::{
+    EnumItem, EnumVariant, Field, FragmentItem, Identifier, Item, KeywordDefinition, KeywordItem,
+    KeywordValue, Language, OperatorModel, PrecedenceExpression, PrecedenceItem,
+    PrecedenceOperator, PrimaryExpression, RepeatedItem, Scanner, SeparatedItem, StructItem,
+    TokenDefinition, TokenItem, TriviaItem, VersionSpecifier,
+};
+use indexmap::IndexMap;
+use inflector::Inflector;
+
+use crate::model::{Definition, DefinitionKind, Entry, Expression, Value};
+
+pub struct Builder {
+    entries: HashMap<String, Entry>,
+}
+
+impl Builder {
+    pub fn build(language: &Language) -> HashMap<String, Entry> {
+        let mut builder = Self {
+            entries: HashMap::new(),
+        };
+
+        for item in language.items() {
+            match item {
+                Item::Struct { item } => builder.build_struct_item(item),
+                Item::Enum { item } => builder.build_enum_item(item),
+                Item::Repeated { item } => builder.build_repeated_item(item),
+                Item::Separated { item } => builder.build_separated_item(item),
+                Item::Precedence { item } => builder.build_precedence_item(item),
+                Item::Trivia { item } => builder.build_trivia_item(item),
+                Item::Keyword { item } => builder.build_keyword_item(item),
+                Item::Token { item } => builder.build_token_item(item),
+                Item::Fragment { item } => builder.build_fragment_item(item),
+            };
+        }
+
+        builder.entries
+    }
+
+    fn add_entry(&mut self, name: &Identifier, is_terminal: bool, is_inlined: bool) {
+        let mut ebnf_id = name.to_string();
+
+        if is_terminal {
+            ebnf_id = ebnf_id.to_screaming_snake_case();
+        }
+
+        if is_inlined {
+            ebnf_id = format!("«{ebnf_id}»");
+        }
+
+        let existing_entry = self.entries.insert(
+            name.to_string(),
+            Entry::new(name.to_string(), ebnf_id, vec![]),
+        );
+
+        assert!(
+            existing_entry.is_none(),
+            "Entry '{name}' is already defined."
+        );
+    }
+
+    fn add_definition(
+        &mut self,
+        name: &Identifier,
+        leading_comments: impl IntoIterator<Item = String>,
+        values: impl IntoIterator<Item = Value>,
+        kind: DefinitionKind,
+    ) {
+        let definition = Definition::new(
+            leading_comments.into_iter().collect(),
+            values.into_iter().collect(),
+            kind,
+        );
```

Comment:
Don't mean to outline this function; it's just that using bare struct construction syntax is simpler and tells that nothing else is going on: 
```rust
let definition = Definition {
  leading_comments: ..,
  value: ..,
  kind,
};
```

Here, `new` only hides what's going on and introduces reliance on the order rather than being explicit. This is not important and is style but wanted to note that.

Reply:
I agree, but I think that is more important in case if was something like `codegen_language_definition::model::*` where it is used everywhere, and being able to grep/understand it across the entire codebase is important, and also it needs to survive renaming/refactoring over time. But in this case, for these internal types that are only instantiated (repeatedly) in one file, the brevity actually helps make the surrounding code more readable IMO, so that it can focus on the logic, rather than the verbose field/value pairs everywhere. I think this benefit (given how frequently these initializations happen here) outweighs that.

Reply:
Agreed, I'm not blocking on that and I mentioned that it's ultimately not important and mostly a style issue. I noticed the addition of `derive-new` so wanted to mention that sometimes it's helpful but sometimes it's actually worse.

At the cost of stutter/repeating in struct init syntax, we know this has no post-process logic, we know it's a POD, can look up fields in IDE on hover etc.

Alas, I may have donned my Rust Evangelism Strikeforce hat and didn't mean to make this into an actual issue, excuse me :see_no_evil:

Reply:
🦀